Add a postprocessor that allows the visualization of the stress.

bangerth opened the pull request:
Also add a test for the strain rate viz postprocessor.

This closes #288.

gassmoeller commented:
I agree that apart from @jdannberg 's comments this is a valuable addition to ASPECT's postprocessors for many purposes. However, I am not sure if #288 is closed by this entirely. This postprocessor calculates the shear stress tensor, whereas if I remember correctly @dsarahstamps was looking for the principal components of the full stress tensor. This would mean adding the pressure to the stress tensor as `sigma + pressure * I` and then calculating and outputting the eigenvectors and eigenvalues of this tensor, which are the axis and values of the principal stress directions. @dsarahstamps is this correct? Or can you do all necessary computations in Paraview and are fine with this state?

dsarahstamps commented:
@gassmoeller You are correct that I am seeking the principal components of the stress tensor in addition to the shear components.

ian-r-rose commented:
Sarah, I have a function that solves for the eigenvectors and eigenvalues
of a symmetric 3x3/2x2 tensor.  You can find it in this file:

https://github.com/ian-r-rose/aspect/blob/centrifugal/source/postprocess/liouville.cc

If people want to, we could try to add it as a utility during the hackathon.

On Thu, Apr 30, 2015 at 7:43 AM, dsarahstamps <notifications@github.com>
wrote:

> @gassmoeller <https://github.com/gassmoeller> You are correct that I am
> seeking the principal components of the stress tensor in addition to the
> shear components.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/aspect/pull/319#issuecomment-97824766>.
>

dsarahstamps commented:
Thanks Ian. I am interested in working on this contribution to ASPECT at the hackathon.

bangerth commented:
Let's take an incremental approach: I'm going to change the name of this postprocessor to "shear stress", create another one I call "stress" that also contains the pressure, and let both output all tensor components. At the workshop we can then clone one or the other and create something that outputs the principal components and vectors. Sounds good? I suppose the plan above already gets us 3/4 of the way to where we want to be.

gassmoeller commented:
Yes, I agree. This postprocessor is still very useful for other purposes. Let me know, when I should have a look through the code.

dsarahstamps commented:
Yes, that sounds great Wolfgang.

gassmoeller commented on include/aspect/postprocess/visualization/shear_stress.h:
```diff
@@ -0,0 +1,96 @@
+
+          /**
+           * Return the vector of strings describing the names of the computed
+           * quantities. Given the purpose of this class, this is a vector with dim
```
I think the number of entries is now `SymmetricTensor<2,dim>::n_independent_components`

gassmoeller commented on include/aspect/postprocess/visualization/shear_stress.h:
```diff
@@ -0,0 +1,96 @@
+           * output files that consist of more than one data set are to be
+           * interpreted. Since the current class is meant to be used for a single
+           * vector result variable, the returned value is obviously
+           * DataComponentInterpretation::component_is_part repeated dim times.
```
same here

gassmoeller commented on include/aspect/postprocess/visualization/stress.h:
```diff
@@ -0,0 +1,95 @@
+           * output files that consist of more than one data set are to be
+           * interpreted. Since the current class is meant to be used for a single
+           * vector result variable, the returned value is obviously
+           * DataComponentInterpretation::component_is_part repeated dim times.
```
see above

gassmoeller commented on include/aspect/postprocess/visualization/stress.h:
```diff
@@ -0,0 +1,95 @@
+
+          /**
+           * Return the vector of strings describing the names of the computed
+           * quantities. Given the purpose of this class, this is a vector with dim
```
another one :-)

gassmoeller commented:
Looks very good. Did you update the test results yet? And the astyle-tester is complaining about a single space. Apart from that I would say it is ready to merge. Thanks for pushing this forward :+1:

tjhei commented:
Yeah, tests/visualization-stress needs to be corrected.

tjhei commented:
Also please rebase to current master.

bangerth commented:
I wasn't actually finished rewriting it all :-) So apologies for the inkomplet test output -- I still needed to add some gravity to the stress test to make sure that there is some pressure. Thanks, though, for all of the very valid corrections!

I think I've addressed it all and rebased. Please take another look.

gassmoeller commented on tests/visualization-stress.prm:
```diff
@@ -0,0 +1,103 @@
+end
+
+
+# no gravity. the pressure will equal just the dynamic component
```
copy-paste from the other tests, but no longer true in this case ;-)

gassmoeller commented:
Apart from the one comment, this looks ready to merge. Please fix and merge, whenever you have the time. Thanks for doing this.
